data/txDupCache.go

```diff
@@ -0,0 +1,228 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package data
+
+import (
+	"context"
+	"encoding/binary"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-deadlock"
+
+	"golang.org/x/crypto/blake2b"
+)
+
+// digestCache is a rotating cache of size N accepting crypto.Digest as a key
+// and keeping up to 2*N elements in memory
+type digestCache struct {
+	cur  map[crypto.Digest]struct{}
+	prev map[crypto.Digest]struct{}
+
+	maxSize int
+	mu      deadlock.Mutex
+}
+
+func makeDigestCache(size int) *digestCache {
+	c := &digestCache{
+		cur:     map[crypto.Digest]struct{}{},
+		maxSize: size,
+	}
+	return c
+}
+
+func (c *digestCache) check(d *crypto.Digest) bool {
+	_, found := c.cur[*d]
+	if !found {
+		_, found = c.prev[*d]
+	}
+	return found
+}
+
+func (c *digestCache) swap() {
+	c.prev = c.cur
+	c.cur = map[crypto.Digest]struct{}{}
+}
+
+func (c *digestCache) put(d *crypto.Digest) {
+	if len(c.cur) >= c.maxSize {
+		c.swap()
+	}
+	c.cur[*d] = struct{}{}
+}
```

Comment:
These are safe because they are only used inside `checkAndPut` which handles the mutex locking but do we have a preferred pattern for handling situations like these in other places? Perhaps putting a comment noting that it assumes that it has the lock and shouldn't be used outside of `checkAndPut`?

Reply:
done